core/src/main/java/org/apache/iceberg/taskio/BaseTaskWriter.java

```diff
@@ -0,0 +1,141 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.taskio;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Supplier;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.DataFiles;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.Metrics;
+import org.apache.iceberg.PartitionKey;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.encryption.EncryptedOutputFile;
+import org.apache.iceberg.io.FileAppender;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.util.Tasks;
+
+abstract class BaseTaskWriter<T> implements TaskWriter<T> {
+  protected static final int ROWS_DIVISOR = 1000;
+
+  private final List<DataFile> completedFiles = Lists.newArrayList();
+  private final PartitionSpec spec;
+  private final FileFormat format;
+  private final FileAppenderFactory<T> appenderFactory;
+  private final OutputFileFactory fileFactory;
+  private final FileIO io;
+  private final long targetFileSize;
+
+  protected BaseTaskWriter(PartitionSpec spec, FileFormat format, FileAppenderFactory<T> appenderFactory,
+                           OutputFileFactory fileFactory, FileIO io, long targetFileSize) {
+    this.spec = spec;
+    this.format = format;
+    this.appenderFactory = appenderFactory;
+    this.fileFactory = fileFactory;
+    this.io = io;
+    this.targetFileSize = targetFileSize;
+  }
+
+  @Override
+  public void abort() throws IOException {
+    close();
+
+    // clean up files created by this writer
+    Tasks.foreach(completedFiles)
+        .throwFailureWhenFinished()
+        .noRetry()
+        .run(file -> io.deleteFile(file.path().toString()));
+  }
+
+  @Override
+  public List<DataFile> pollCompleteFiles() {
+    if (completedFiles.size() > 0) {
+      List<DataFile> dataFiles = ImmutableList.copyOf(completedFiles);
+      completedFiles.clear();
+      return dataFiles;
+    } else {
+      return Collections.emptyList();
+    }
+  }
+
+  protected OutputFileFactory outputFileFactory() {
+    return this.fileFactory;
+  }
+
+  WrappedFileAppender createWrappedFileAppender(PartitionKey partitionKey,
+                                                Supplier<EncryptedOutputFile> outputFileSupplier) {
+    EncryptedOutputFile outputFile = outputFileSupplier.get();
+    FileAppender<T> appender = appenderFactory.newAppender(outputFile.encryptingOutputFile(), format);
+    return new WrappedFileAppender(partitionKey, outputFile, appender);
+  }
+
+  class WrappedFileAppender {
```

Comment:
I don't see much value in this class. Its primary use is to keep track of whether a file is large enough to release, but it doesn't actually have any of the logic to do that. As a consequence, the code is now split across multiple places.

This also has the logic for closing an appender and converting it to a `DataFile`, but that could just as easily be done in a `DataFile closeAppender(FileAppender appender)` method.

It would make sense to keep this class if it completely encapsulated the logic of rolling new files. That would require some refactoring so that it could create new files using the file and appender factories. It would also require passing a `Consumer<DataFile>` so that it can release closed files. Otherwise, I think we should remove this class.

Reply:
I created this class because in fanout writer  we will have several opening writers  and when building the `DataFile`,  we will need all the informations for the given FileAppender,  such as partitionKey,  EncryptedOutputFile etc. The previous spark implementations won't need the class because all of the context information are maintained inside the  PartitionedWriter (currentXXX ), that's not work for fanout writer. It will be better to have such a class to hold those infos to build `DataFile`. 

> It would make sense to keep this class if it completely encapsulated the logic of rolling new files
Good point.   Make the WrappedFileAppender to accomplish all the rolling things,  let me refactor this.